### ISSUE TITLE
Fluent Api ignores supplied propertyName

### ISSUE THREAD

**Issue by dpwzone** [dotnet/efcore#31813]
When using the EF core Fluent Api to construct a many to many relationship, I supply it both the relatedTypeName and a propertyName in "HasMany".  However, when looking at the ModelBuilder.Model.DebugView, I see my newly created SkipNavigation is named "Item", instead of the supplied propertyName....  why is it "Item"??!?  Why even give an option to supply the propertyName, if you're just going to ignore it.  That's just incredibly poor programming!

Now I have to write a lot of extra code to clean up something that shouldn't have been a problem to start with, because there is next to no documentation, and I'm getting tired of hitting dead ends.

**Comment by ajcvickers** [comment]
This issue is lacking enough information for us to be able to fully understand what is happening. Please attach a small, runnable project or post a small, runnable code listing that reproduces what you are seeing so that we can investigate.

**Comment by dpwzone** [comment]
Given the size and complexity of our codebase, it would be difficult to pull out code, but the following screenshots should provide more than enough detail.

In this screenshot, I'm starting to add the one-to-many relationship using the fluent api.  So far, the modelBuilder.Model looks good:
![image](https://github.com/dotnet/efcore/assets/100529842/f22eabb2-b574-43d7-8e4b-26e94ff3d02d)

In this screenshot, I just added the one-to-many, but the modelBuilder.Model shows "Item" instead of "Make":
![image](https://github.com/dotnet/efcore/assets/100529842/46ea1ae7-ecde-46d6-9a4b-9a13d30883b1)

In this screenshot, you can see the signature specifically mentions "navigationName"...  why isn't this being used by EF?  Usually when a program allows for user input, it uses it.  It doesn't just say, whoops, I know better than the user.
![image](https://github.com/dotnet/efcore/assets/100529842/66d7fe96-25c0-4e43-a599-8bd02cbbe61b)

**Comment by ajcvickers** [comment]
@dpwzone There isn't enough here to reproduce what you are seeing. If you can't put it into a small standalone application with all the knowledge of the actual code, then the chances of us being able to do so are very small.

**Comment by dpwzone** [comment]
@ajcvickers Okay, figured that would be the response.  Let's try a different route then...  I take it that it's the internal ICandidateNamingService generating the names, is that a correct assumption?  If so, how do I override that service?  Adding it to the application services and/or the dbContext services doesn't do anything.

I tried adding my own CandidateNamingService singleton...  that didn't work, apparently EF doesn't care about this.

I tried the route of using the AddEntityFrameworkDesignTimeServices()...  that didn't work, threw an error about static vs scoped logger.

I tried the route of IDesignTimeServices...  that didn't work, my class is never called.


Honestly, it's poor programming that the "navigationName" isn't used....  what's it there for, if not used???  That's like me presenting a form for a user to fill in their first and last names, and then me going, nope your name is actually John Smith.

**Comment by ajcvickers** [comment]
> Honestly, it's poor programming that the "navigationName" isn't used.... w

It is used. Apparently in your case, it at least is not showing up in the model at that point, which may or may not be a bug. We need to be able to reproduce it in order to fix it or explain why it behaves the way it does. There's no point in my providing guidance for overriding services or the like without understanding what is going on in the first place.

**Comment by dpwzone** [comment]
> It is used.

Where?  What class/method handles that?

> There's no point in my providing guidance for overriding services or the like without understanding what is going on in the first place.

I don't think that's your call.  I'm a programmer trying to learn more about how to use your product.  If you don't know how to override it, just say that.

So, is EF using the CandidateNamingService, and if so, why is my custom CandidateNamingService not ever called?

I added it via:  builder.Services.AddSingleton<ICandidateNamingService, MyCandidateNamingService>();  in program.cs;

**Comment by dpwzone** [comment]
Does anyone involved on this project actually know how things are named?  Seems like a simple ask, yet crickets...  but then again, your programmers don't think things should be named as what you input.

I've got navigation properties named "Item" that EF Core added....  surely you all know how your navigations are added, and why something would be named "Item"

**Comment by roji** [comment]
@dpwzone I've also gone over the above exchange, and we really will need a small code sample, as @ajcvickers requested. I understand the urge to ask questions, and also that putting together a repro may seem daunting or time-consuming, but it really is the easiest forward both for you and for us. If you're supplying the property name, then that definitely **should** get used; if it isn't, that may be a bug in EF, or it may be an issue with how your code is using EF Core's APIs. The only way of knowing what's going on is a minimal, runnable program that shows the problem occuring.

Note that we don't need (or want) to see your actual project: ideally you'd start a new console program, and do the minimal code needed to reproduce the error. In the majority of cases, as users attempt to prepare such a program, they figure out the error in their own code (e.g. since they fail to reproduce the error in isolation). If you manage to reproduce the error in such a minimal program, that's the perfect way for us to investigate exactly what's going on.

I'd also recommend against prematurely diving into EF's internals (e.g. ICandidateNamingService) and trying to figure out what's going on; that's time better spent on creating the minimal code sample. Once we understand what's going on, we'd be happy to explain the problem (or to fix any bugs, if any are found).

Finally, please allow us some time when waiting for a response - we're a small team with a pretty load of incoming Github issues (not to mention other work). It sometimes takes us a while to post replies.

**Comment by dpwzone** [comment]
Look, telling a programmer to supply their code is weak.  Yes, it makes it easier for you all, but puts the burden on the programmer having issues.

Do you, yes or no, know how navigation names are set?  If yes, surely you'd know why a navigation of "Item" is created.  It sounds like a default to me, but that's why I'm asking you all.

**Comment by roji** [comment]
> Look, telling a programmer to supply their code is weak. Yes, it makes it easier for you all, but puts the burden on the programmer having issues.

Believe me, we manage support for a massive number of users on a regular basis, and we're not doing this to make our lives easier. The time users themselves lose on various back-and-forths where they try to describe their problems to us is considerably more than it would take to simply produce a code sample. This isn't weak, it's just the way software engieering works - go to any project out there with a problem, and they'll ask for a repro.

So to summarize, we do know how navigation names are set; this setting works perfectly well for the vast majority of our users. This indicates you're doing something special or non-standard, which is why we need to see what you're doing.

**Comment by dpwzone** [comment]
So to summarize,  you're not going to explain how navigation names are set...  got it.

Hate to break it to you, but if your code is returning a navigation called "Item", and not what I named it, then that's a bug in YOUR code, not mine.

HasOne takes the related type entity and navigation name...  the navigation name is being ignored by YOUR code.

So, get off your high horse, and try looking to see what this bug could possibly be...  why would your code produce a navigation called "Item"??

**Comment by roji** [comment]
@dpwzone I'm closing this as a duplicate of #31836 since there doesn't seem to be anything here that's different from that issue.

If you choose to provide a minimal repro - as is standard and as is done by the vast majority of our users - we'll be happy to investigate what's going on. Otherwise this isn't a forum for explaining EF's entire internal architecture to all our users.

**Comment by roji** [comment]
Duplicate of #31836

**Comment by dpwzone** [comment]
What's the matter, you afraid to show you don't know how these navigation names are set? How about leaving this issue open for someone who does?

**Comment by roji** [comment]
https://github.com/dotnet/efcore/issues/31836#issuecomment-1732076931

**Comment by dpwzone** [comment]
Cool, keep thinking you're good at what you do.  Meanwhile, I'll recommend to the fortune 500 company that I work for, to drop EF because they're bug trackers are incapable.
